Use 10x cellranger for initial fastq generation

```bash
#set up environment variables 
export SCRATCH="/volumes/USR2/Ryan/work"
export projDir="/volumes/USR2/Ryan/projects/10x_MET"
export srcDir="${projDir}/src"
export refDir="/volumes/USR2/Ryan/ref/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"
export outDir="${projDir}/241105_10xmet_231"
export DNArunDir="/volumes/seq/flowcells/MDA/nextseq2000/2024/241103_RM_10xmet_atrandi_met/241103_VH01788_59_AAG2N3JM5"
export outname="10xmet_231"
export task_cpus=250
export index="/volumes/USR2/Ryan/projects/metact/ref/bsbolt/genome_index.pkl"

mkdir -p $outDir
cd $outDir

echo """Lane,Sample,Index
*,${outname}_vera_noconv,SI-NA-G1
*,${outname}_vera_conv,SI-NA-G3
*,${outname}_dmet_noconv,SI-NA-G2
*,${outname}_dmet_conv,SI-NA-G4""" > sample_met.csv
export sample_met="${outDir}/sample_met.csv"

mkdir -p ${SCRATCH}/10xmet_work

~/tools/cellranger-atac-2.1.0/cellranger-atac mkfastq \
--run=${DNArunDir} \
--csv=${sample_met} \
--jobmode="local" \
--localcores=${task_cpus} \
--delete-undetermined \
--output-dir="${outDir}" \
--id=${outname}_met

```

Set up fastq files with indexes for alignment

```python
import gzip
from Bio import SeqIO
import sys
from Bio.SeqIO.QualityIO import FastqGeneralIterator
from Bio.Seq import Seq
import pandas as pd
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--fq1',default="10xmet_231_vera_conv_S2_L001_R1_001.fastq.gz")
parser.add_argument('--fq2',default="10xmet_231_vera_conv_S2_L001_R3_001.fastq.gz")
parser.add_argument('--idx3',default="10xmet_231_vera_conv_S2_L001_I1_001.fastq.gz")
parser.add_argument('--idx4',default="10xmet_231_vera_conv_S2_L001_R2_001.fastq.gz")
args = parser.parse_args()

fq1=args.fq1
fq2=args.fq2
idx3=args.idx3
idx4=args.idx4

i=0

#open fastq files, correct barcode read names then out fastq 1 and 2  with new read name
with gzip.open(fq1, "rt") as handle1, \
     gzip.open(fq2, "rt") as handle2, \
     gzip.open(idx3, "rt") as handle3, \
     gzip.open(idx4, "rt") as handle4, \
     open(fq1[:-9]+".met.barc.fastq", "w") as outfile_fq1, \
     open(fq2[:-9]+".met.barc.fastq", "w") as outfile_fq2:
        for (title1, seq1, qual1), \
        (title2, seq2, qual2), \
        (title3,seq3,qual3), \
        (title4,seq4,qual4) in \
        zip(FastqGeneralIterator(handle1), \
                FastqGeneralIterator(handle2),\
                FastqGeneralIterator(handle3),\
                FastqGeneralIterator(handle4)):
                        sample_idx=seq3[:10]
                        gem_idx=seq4[:16]
                        cell_idx=sample_idx+"_"+gem_idx
                        outfile_fq1.write("@%s:%s\n%s\n+\n%s\n" % (cell_idx, i, seq1, qual1))
                        outfile_fq2.write("@%s:%s\n%s\n+\n%s\n" % (cell_idx, i, seq2, qual2))
                        i+=1
```

# Run fastq barcode generator into readnames

```bash
cd ${outDir}/AAG2N3JM5
singularity shell ~/singularity/amethyst.sif

python $srcDir/fastqsplitter.10xmet.py \
--fq1 10xmet_231_dmet_conv/10xmet_231_dmet_conv_S4_L001_R1_001.fastq.gz \
--fq2 10xmet_231_dmet_conv/10xmet_231_dmet_conv_S4_L001_R3_001.fastq.gz \
--idx3 10xmet_231_dmet_conv/10xmet_231_dmet_conv_S4_L001_I1_001.fastq.gz \
--idx4 10xmet_231_dmet_conv/10xmet_231_dmet_conv_S4_L001_R2_001.fastq.gz &

python $srcDir/fastqsplitter.10xmet.py \
--fq1 10xmet_231_dmet_noconv/10xmet_231_dmet_noconv_S3_L001_R1_001.fastq.gz \
--fq2 10xmet_231_dmet_noconv/10xmet_231_dmet_noconv_S3_L001_R3_001.fastq.gz \
--idx3 10xmet_231_dmet_noconv/10xmet_231_dmet_noconv_S3_L001_I1_001.fastq.gz \
--idx4 10xmet_231_dmet_noconv/10xmet_231_dmet_noconv_S3_L001_R2_001.fastq.gz &

python $srcDir/fastqsplitter.10xmet.py \
--fq1 10xmet_231_vera_conv/10xmet_231_vera_conv_S2_L001_R1_001.fastq.gz \
--fq2 10xmet_231_vera_conv/10xmet_231_vera_conv_S2_L001_R3_001.fastq.gz \
--idx3 10xmet_231_vera_conv/10xmet_231_vera_conv_S2_L001_I1_001.fastq.gz \
--idx4 10xmet_231_vera_conv/10xmet_231_vera_conv_S2_L001_R2_001.fastq.gz &

python $srcDir/fastqsplitter.10xmet.py \
--fq1 10xmet_231_vera_noconv/10xmet_231_vera_noconv_S1_L001_R1_001.fastq.gz \
--fq2 10xmet_231_vera_noconv/10xmet_231_vera_noconv_S1_L001_R3_001.fastq.gz \
--idx3 10xmet_231_vera_noconv/10xmet_231_vera_noconv_S1_L001_I1_001.fastq.gz \
--idx4 10xmet_231_vera_noconv/10xmet_231_vera_noconv_S1_L001_R2_001.fastq.gz &

#when they are all done
gzip ${outDir}/AAG2N3JM5/*/*barc.fastq

```
# Run adapter trimming and alignment
```bash
singularity shell --bind ~/projects ~/singularity/scalemethyl_v1.6.sif
cd ${outDir}/AAG2N3JM5
# run fastqc to assess conversion rate
fastqc -t 100 ./*/*fastq.gz
parallel -j 8 gzip ::: $(ls ./*/*met.barc.fastq)
#from fastqc we can see that genomic dna is converted compared to nonconverted strand! (yay!) and we need to trim some adapters


#run demultiplexing
adapter_trim() { 
cutadapt \
-j 50 \
-a CTGTCTCTTA -A CTGTCTCTTA \
-U 10 -u 10 \
-o $(ls ${1}/${1}*_L001_R1_001.met.barc.fastq.gz) \
-p $(ls ${1}/${1}*_L001_R3_001.met.barc.fastq.gz)
}

export -f adapter_trim

adapter_trim 10xmet_231_dmet_conv &
adapter_trim 10xmet_231_dmet_noconv & 
adapter_trim 10xmet_231_vera_conv &
adapter_trim 10xmet_231_vera_noconv &

align() {
bsbolt Align \
-F1 ${pairs[1]} -F2 ${pairs[0]} \
-t 50 -OT 50 \
-O $sample \
-DB $index >> ${sample}.bsbolt.log 2>> ${sample}.bsbolt.log
}

export -f align

align 10xmet_231_dmet_conv &
align 10xmet_231_dmet_noconv & 
align 10xmet_231_vera_conv &
align 10xmet_231_vera_noconv &
