Experiment: https://mdandersonorg-my.sharepoint.com/personal/rmulqueen_mdanderson_org/_layouts/OneNote.aspx?id=%2Fpersonal%2Frmulqueen_mdanderson_org%2FDocuments%2FmetACT&wd=target%2810xmet.one%7C6C1BDECA-4977-0A4C-AD58-1D18D92E3B99%2F241211%2010xWGS%20ATAC%20Kit%7C69BFB731-8F8E-EE4A-AD35-37A0BFA2E4E6%2F%29
onenote:https://mdandersonorg-my.sharepoint.com/personal/rmulqueen_mdanderson_org/Documents/metACT/10xmet.one#241211%2010xWGS%20ATAC%20Kit&section-id={6C1BDECA-4977-0A4C-AD58-1D18D92E3B99}&page-id={69BFB731-8F8E-EE4A-AD35-37A0BFA2E4E6}&end


Sequencing folder: /volumes/seq/flowcells/MDA/nextseq2000/2024/20241218_RM_10xWGS_DCIS676T

Ran 10X ATAC kit after Nucleosome depletion on 300k input viable dissociated cells for DCIS sample BCMDCIS66T. 

Use 10x cellranger for initial fastq generation

```bash
##################NOT RUN#####################

#set up environment variables 
export DNArunDir="/volumes/seq/flowcells/MDA/nextseq2000/2024/20241218_RM_10xWGS_DCIS676T"
export experiment_name="241218_10xWGS_DCIS676T"
export SCRATCH="/volumes/USR2/Ryan/work"
export projDir="/volumes/USR2/Ryan/projects/10x_MET"
export srcDir="${projDir}/src"
export outDir="${projDir}/${experiment_name}"
export outname="${experiment_name}"
export refDir="/volumes/USR2/Ryan/ref/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"
export index="/volumes/USR2/Ryan/projects/metact/ref/bsbolt/genome_index.pkl"
export task_cpus=250

mkdir -p $outDir
cd $outDir

echo """Lane,Sample,Index
*,${experiment_name},SI-NA-B5""" > sample_met.csv
export sample_met="${outDir}/sample_met.csv"

mkdir -p ${SCRATCH}

~/tools/cellranger-atac-2.1.0/cellranger-atac mkfastq \
--run=${DNArunDir} \
--csv=${sample_met} \
--jobmode="local" \
--localcores=${task_cpus} \
--delete-undetermined \
--output-dir="${outDir}" \
--id=${outname}

#can use regular cell ranger for this
~/tools/cellranger-atac-2.1.0/cellranger-atac count \
                        --id=${outname} \
                        --reference=${refDir} \
                        --fastqs="${outDir}/${outname}/outs/fastq_path/241218_10xWGS_DCIS676T/outs/fastq_path/2225WKTNX/241218_10xWGS_DCIS676T" \
                        --localcores=100 \
                        --project "${outname}" \
                        --localmem=200

##################NOT RUN#####################
#split out single cells to estimate complexity

singularity shell \
--bind /volumes/USR2/Ryan/projects/metact/ref/:/ref \
~/singularity/amethyst.sif
source ~/.bashrc

mkdir -p ${outDir}/sc_bam
bam_in="/volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T/2225WKTNX/241218_10xWGS_DCIS676T/241218_10xWGS_DCIS676T/outs/possorted_bam.bam"

samtools split -d CB \
-u ${outDir}/unassigned_reads.bam \
-M 6000 --output-fmt BAM \
-@ 200 \
-f ${outDir}'/sc_bam/%*_%!.%.' ${bam_in}

#remove files with size less than 10MB
find . -maxdepth 1 -type f -name "*.bam" -size -10000k -delete

#* Based on the Lander-Waterman equation that states:
#*     C/X = 1 - exp( -N/X )
#* where
#*     X = number of distinct molecules in library
#*     N = number of read pairs
#*     C = number of distinct fragments observed in read pairs
cd /volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T/sc_bam
mkdir -p ./bbrd_bam
mkdir -p ./sc_stats
mkdir -p ./sorted_se_bam

lib_complexity(){
    bam=$1
    samtools sort -n $bam | samtools fixmate -p -m - - | samtools sort | samtools markdup -S - ./bbrd_bam/${bam::-4}.bbrd.bam
    java -jar ~/tools/picard.jar EstimateLibraryComplexity \
        I="./bbrd_bam/${bam::-4}.bbrd.bam" \
        O="./sc_stats/${bam::-4}.metrics.txt"
    samtools sort -S - ./bbrd_bam/${bam::-4}.bbrd.bam
    }

export -f lib_complexity
parallel -j 200 lib_complexity ::: $(ls *-1.bam)

for i in *metrics.txt; 
do echo -e "$i \t $(grep "^MissingLibrary.1" $i | awk 'OFS="\t" {print $3,$9,$10}')"; 
done > ${outDir}/read_counts.tsv

cd /volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T/sc_bam/bbrd_bam
pe_to_se(){
    bam=$1
    samtools sort -o ../sorted_se_bam/${bam::-4}.se.bam $bam 
}
export -f pe_to_se
parallel -j 200 pe_to_se ::: $(ls *bbrd.bam)

```

```R

library(ggplot2)
setwd("/volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T")
dat<-read.table("read_counts.tsv",sep="\t")
colnames(dat)<-c("cellID","total_reads","duplication_rate","estimated_libsize")
dat$duplication_perc<-dat$duplication_rate*100
dat$uniq_perc<-100-dat$duplication_perc

plt<-ggplot(dat,aes(y=log10(total_reads),x=uniq_perc))+geom_point()+theme_minimal()+coord_cartesian(xlim=c(0,100),ylim=c(0,8))
ggsave(plt,file="complexity_plot.pdf")

plt<-ggplot(dat,aes(y=log10(estimated_libsize),x=1))+geom_violin()+geom_jitter()+theme_minimal()+coord_cartesian(ylim=c(0,8))
ggsave(plt,file="estimated_size.pdf")

```

```bash
singularity shell \
--bind ~/projects \
~/singularity/copykit.sif
source ~/.bashrc
```

```R
library(copykit)
library(BiocParallel)
library(optparse)
library(circlize)
BiocParallel::bpparam()

option_list = list(
  make_option(c("-i", "--input_dir"), type="character", default=".", 
              help="List of single-cell bam files", metavar="character"),
  make_option(c("-p", "--output_prefix"), type="character", default="allcells", 
              help="Prefix of output"),
  make_option(c("-c", "--task_cpus"), type="integer", default=1, 
              help="Integer number of cpus")
);

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
cpu_count=100
prefix="dcis66t"

register(MulticoreParam(progressbar = T, workers = cpu_count), default = T)

dat <- runVarbin("/volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T/sc_bam/sorted_se_bam",
    remove_Y = TRUE, genome="hg38",is_paired_end=TRUE,method="multipcf",gamma=20)


dat  <- runMetrics(dat)

#dat <- dat[,colData(dat)$reads_total > 100000]
setwd("/volumes/USR2/Ryan/projects/10x_MET/241218_10xWGS_DCIS676T/")

pdf(paste0(prefix,".qc_metrics.pdf"))
plotMetrics(dat, metric = c("overdispersion", 
                              "breakpoint_count",
                              "reads_total",
                              "reads_duplicates",
                              "reads_assigned_bins",
                              "percentage_duplicates"),
            label = "reads_total")
dev.off()

dat <- dat[,colData(dat)$reads_assigned_bins > 200000 ]
# Mark euploid cells if they exist
dat <- findAneuploidCells(dat,resolution=0.8)

# Mark low-quality cells for filtering
#dat <- findOutliers(dat)
#pdf(paste0(prefix,".outlier_qc.heatmap.pdf"))
#plotHeatmap(dat, label = c('outlier', 'is_aneuploid'), row_split = 'outlier')
#dev.off()

# kNN smooth profiles
dat <- knnSmooth(dat,k=20)

# Create a umap embedding 
dat <- runUmap(dat)
k_clones<-findSuggestedK(dat) #10

dat  <- findClusters(dat, k_superclones=k_clones@metadata$suggestedK-3, k_subclones=k_clones@metadata$suggestedK+3)#output 

# Calculate consensus profiles for each subclone, 
# and order cells by cluster for visualization with plotHeatmap
dat <- calcConsensus(dat)
dat <- runConsensusPhylo(dat)
dat <- runPhylo(dat, metric = 'manhattan')
dat <- calcInteger(dat, method = 'scquantum', assay = 'smoothed_bincounts')
pdf("cell_ploidy.pdf")
plotMetrics(dat, metric = 'ploidy', label = 'ploidy_score')
dev.off()

dat <- dat[,colData(dat)$ploidy_score < 0.25]
dat <- calcInteger(dat, method = 'fixed', ploidy_value = 2)
dat <- calcConsensus(dat, consensus_by = 'subclones', assay = 'integer')

# Plot a copy number heatmap with clustering annotation
pdf(paste0(prefix,".subclone.umap.pdf"))
plotUmap(dat,label="subclones")
dev.off()

pdf(paste0(prefix,".subclone.heatmap_int.pdf"))
plotHeatmap(dat, label = c('reads_total','subclones'),
    assay = 'integer',
    order='consensus_tree',
    col=colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
    n_threads=50)
dev.off()

pdf(paste0(prefix,".subclone.heatmap_seg.pdf"))
plotHeatmap(dat, label = c('reads_total','subclones'),
    order='consensus_tree',
    n_threads=50,
    col=colorRamp2(c(-2,-1, 0, 1,2), c("darkblue","blue", "white", "red","darkred")))
dev.off()


pdf(paste0(prefix,".subclone.heatmap_seg.pdf"))
plotHeatmap(dat, label = c('reads_total','subclones'),
    order='consensus_tree',
    n_threads=50,
    col=colorRamp2(c(-1, 0, 1), c("darkblue", "white", "darkred")))
dev.off()


pdf(paste0(prefix,".subclone.phylo.pdf"))
plotPhylo(dat, label = 'subclones')
dev.off()

pdf(paste0(prefix,".subclone.consensus.pdf"))
plotHeatmap(dat,
            consensus = TRUE,
            assay = 'integer',
            label = 'subclones')
dev.off()

saveRDS(dat,file=paste0(prefix,".scCNA.rds"))
write.table(as.data.frame(dat@colData),file=paste0(prefix,".scCNA.tsv"),sep="\t",col.names=T,row.names=T)
```